static/src/javascripts/projects/common/modules/commercial/dfp/messenger.js

```diff
@@ -0,0 +1,149 @@
+define([
+    'Promise',
+    'common/utils/report-error'
+], function (Promise, reportError) {
+    var currentHost = location.protocol + '//' + location.host;
+    var listeners = {};
+    var registeredListeners = 0;
+
+    var error400 = { code: 400, message: 'Bad request' };
+    var error405 = { code: 405, message: 'Service %% not implemented' };
+    var error418 = { code: 418, message: 'I\'m a teapot' };
+    var error500 = { code: 500, message: 'Internal server error\n\n%%' };
+
+    return {
+        register: register,
+        unregister: unregister
+    };
+
+    function register(type, callback, _window) {
+        if( registeredListeners === 0 ) {
+            on(_window || window);
+        }
+
+        listeners[type] || (listeners[type] = []);
+        if (listeners[type].indexOf(callback) === -1) {
+            listeners[type].push(callback);
+            registeredListeners += 1;
+        }
+    }
+
+    function unregister(type, callback, _window) {
+        if (callback === undefined) {
+            registeredListeners -= listeners[type].length;
+            listeners[type].length = 0;
+        } else {
+            var idx = listeners[type].indexOf(callback);
+            if (idx > -1) {
+                registeredListeners -= 1;
+                listeners[type].splice(idx, 1);
+            }
+        }
+
+        if (registeredListeners === 0) {
+            off(_window || window);
+        }
+    }
+
+    function on(window) {
+        window.addEventListener('message', onMessage);
+    }
+
+    function off(window) {
+        window.removeEventListener('message', onMessage);
+    }
+
+    function onMessage(event) {
+        // We only allow communication with ads created by DFP
+        if (event.origin !== currentHost) {
+            respond(error400, null);
+            return;
+        }
+
+        try {
+            // Even though the postMessage API allows passing objects as-is, the
+            // serialisation/deserialisation is slower than using JSON
+            // Source: https://bugs.chromium.org/p/chromium/issues/detail?id=536620#c11
+            var data = JSON.parse(event.data);
+        } catch( ex ) {
+            respond(error418, null);
+            return;
+        }
+
+        if (!isValidPayload(data)) {
+            respond(error400, null);
+            return;
+        }
+
+        // If there is no routine attached to this event type, we just answer
+        // with an error code
+        if (!listeners[data.type].length) {
+            respond(formatError(error405, data.type), null);
+            return;
+        }
+
+        var promise = listeners[data.type]
```

Comment:
on a second look, this is slightly dangerous because you're calling 

```js
messenger.register('set-ad-height', function onSetAdHeight(data) {
    if (maybe) {
        messenger.unregister('set-ad-height', onSetAdHeight);
    }
});
```
that is modifying the listeners array while iterating on it.


It looks like this code
```js
[1,2,3].map(function (value, position, array) {
   console.log('val', value);
   if (position === 0) { array.splice(position, 1) }
})
val 1
val 3
```
as you can see, the value `2` is skipped.

What saves you is that you're wrapping everything in a `Promise`, very sensible. Promises are resolved in the next tick.
However that depends on the implementation and considering we changed polyfill already in the past, you might end up having random behaviours in the future.

Usually in these cases I clone the array and iterate on that
`listeners[data.type].slice().map`